### PR TITLE
Prevent Nipype from emitting .proc-* files to user working dir Fix #1955 

### LINKED ISSUE #1955: 🐛 .proc-* files get saved in current working directory
### Describe the bug

.proc-####_time-* files, which are generated by nipype, get saved into the current working directory. It might be deleted after the runs are done, but if there is a crash or if you cancel the job, the files do not get deleted. 

### To reproduce

1. Run C-PAC develop branch
2. You will notice that in the current working directory (wherever you ran the cpac command), the `.proc*` files will appear.

### Preconfig

- [X] default
- [X] abcd-options
- [X] anat-only
- [X] blank
- [X] ccs-options
- [X] fmriprep-options
- [X] fx-options
- [X] monkey
- [X] monkey-ABCD
- [X] ndmg
- [X] nhp-macaque
- [X] preproc
- [X] rbc-options
- [X] rodent

### Custom pipeline configuration

_No response_

### Run command

Tested in ACCESS with singularity: 

```
#!/usr/bin/bash
#SBATCH -N 1
#SBATCH -p RM-shared
#SBATCH -t 14:00:00
#SBATCH --ntasks-per-node=20

HOME=${HOME} \
singularity run \
    --cleanenv \
    -B ${MED} \
    -B ${DATA}:${DATA} \
    -B ${OUTPUT}:${OUTPUT} \
    -B ${CONFIG}:${CONFIG} \
    -B ${PIPELINE_CONFIGS}:${PIPELINE_CONFIGS} \
    ${IMAGE} ${DATA} ${OUTPUT} participant \
    --save_working_dir --skip_bids_validator \
    --pipeline_file ${PIPELINE_CONFIGS}/${pipeline}_seed.yml \
    --data_config_file ${CONFIG}/data_config_regtest_${data}.yml \
    --n_cpus 18 --mem_gb 60
```

### Expected behavior

.proc files get generated and saved in cwd of terminal as soon as CPAC run starts.

### Acceptance criteria

no `.proc*` should be getting written/saved in the current working directory. Maybe they can get moved to a `tmp` folder in the container.

### Screenshots

Screenshot 2023-06-21 at 3 36 33 PM

### C-PAC version

v1.8.6dev

### Container platform

_No response_

### Docker and/or Singularity version(s)

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by nx10**
## Fixes

Fixes #1955 by @amygutierrez

## Description

- Since Nipype version ~1.8.x Nipype emits many (!) `.proc-*` to the _user_ working directory (wherever the user started nipype from, instead of _nipype_ working directory node subdirectories)

## Technical details

@shnizzedy 's exhaustive internal writedown 🕵️
> C-PAC and fMRIPrep both hit the issue upgrading from nipype 1.5.x to 1.8.x; I think [this](https://github.com/nipy/nipype/blob/071523a5069b52689ed27f995f1e2bea5b6e0402/nipype/interfaces/base/core.py#L370) chdir that used to be before [initializing the ResourceMonitor](https://github.com/nipy/nipype/blob/071523a5069b52689ed27f995f1e2bea5b6e0402/nipype/interfaces/base/core.py#L410) led to [this relative path](https://github.com/nipy/nipype/blob/3b505327f412a2c4215d0763f932d0eac6e3a11b/nipype/utils/profiler.py#L59) being inside the relevant subdirectory of the working directory, but now [it's initialized](https://github.com/nipy/nipype/blob/3b505327f412a2c4215d0763f932d0eac6e3a11b/nipype/interfaces/base/core.py#L379-L384) without changing directories first

- The relevant nipype `BaseInterface` is now monkey patched to use the individual node workdir to emit `.proc-*` files to
- This means they will automatically be cleaned if the working dir is not retained (no `--save-work-dir`)

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the `develop` branch of the repository. 
- [x] My commit messages follow best practices.
- [x] My code follows the established code style of the repository.
- [ ] I added tests for the changes I made (if applicable).
- [ ] I updated the changelog.
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: CPAC/pipeline/nipype_pipeline_engine/monkeypatch.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+            with indirectory(cwd or os.getcwd()):  # This is the only line that changed
+                rtc = RuntimeContext(
+                    resource_monitor=config.resource_monitor and self.resource_monitor,
+                    ignore_exception=ignore_exception
+                    if ignore_exception is not None
+                    else self.ignore_exception,
+                )
+
+            with indirectory(cwd or os.getcwd()):
+                self.inputs.trait_set(**inputs)
```
Comment (shnizzedy) [inline CPAC/pipeline/nipype_pipeline_engine/monkeypatch.py:30]: Any reason not to combine these two `with` blocks?

```suggestion
            with indirectory(cwd or os.getcwd()):
                # moved rtc init into context
                rtc = RuntimeContext(
                    resource_monitor=config.resource_monitor and self.resource_monitor,
                    ignore_exception=ignore_exception
                    if ignore_exception is not None
                    else self.ignore_exception,
                )
                self.inputs.trait_set(**inputs)
```

**Review by shnizzedy — APPROVED** [review]
Love this solution! And I think it would be great to offer a non-monkeypatch PR to Nipype itself for this one

**Comment by shnizzedy** [comment]
> This means they will automatically be cleaned if the working dir is not retained (no --save-work-dir)

and that they'll be in their respective subdirectories of the working directory if the working dir is retained!

File: CPAC/pipeline/nipype_pipeline_engine/monkeypatch.py (hunk 1 above)
Comment (nx10) [reply CPAC/pipeline/nipype_pipeline_engine/monkeypatch.py:30]: Yeah this way I could comment `# this is the only line that changed` and future diffs would be marginally simpler. I also did not want to think about if some overloaded constructor might have the ability to change the working dir and it needs to be reset again (probably not. Either way I have no hard feelings

**Pull request merged**